consensus/ibft/fork/manager.go

```diff
@@ -0,0 +1,381 @@
+		return err
+	}
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeSigners() error {
+	for _, fork := range m.forks {
+		valType := fork.ValidatorType
+
+		if err := m.initializeSigner(valType); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeValidatorSets() error {
+	for _, fork := range m.forks {
+		sourceType := ibftTypeToSourceType(fork.Type)
+		if err := m.initializeValidatorSet(sourceType); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeSigner(valType validators.ValidatorType) error {
+	if _, ok := m.signers[valType]; ok {
+		return nil
+	}
+
+	signer, err := signer.NewSignerFromType(m.secretsManager, valType)
+	if err != nil {
+		return err
+	}
+
+	m.signers[valType] = signer
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeValidatorSet(setType valset.SourceType) error {
+	if _, ok := m.validatorSets[setType]; ok {
+		return nil
+	}
+
+	var (
+		valSet valset.ValidatorSet
+		err    error
+	)
+
+	switch setType {
+	case valset.Snapshot:
+		valSet, err = m.initializeSnapshotValidatorSet()
+	case valset.Contract:
+		valSet, err = m.initializeContractValidatorSet()
+	}
+
+	if err != nil {
+		return err
+	}
+
+	m.validatorSets[setType] = valSet
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeSnapshotValidatorSet() (valset.ValidatorSet, error) {
+	snapshotMeta := loadSnapshotMetadata(m.logger, filepath.Join(m.filePath, snapshotMetadataFilename))
+	snapshots := loadSnapshots(m.logger, filepath.Join(m.filePath, snapshotSnapshotsFilename))
+
+	snapshotValset, err := snapshot.NewSnapshotValidatorSet(
+		m.logger,
+		m.blockchain,
+		m.GetSigner,
+		m.epochSize,
+		snapshotMeta,
+		snapshots,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return snapshotValset, nil
+}
+
+func (m *forkManagerImpl) closeSnapshotValidatorSet() error {
+	snapshotValset, ok := m.validatorSets[valset.Snapshot].(*snapshot.SnapshotValidatorSet)
+	if !ok {
+		// no snapshot validator set, skip
+		return nil
+	}
+
+	// save data
+	var (
+		metadata  = snapshotValset.GetSnapshotMetadata()
+		snapshots = snapshotValset.GetSnapshots()
+	)
+
+	if err := writeDataStore(filepath.Join(m.filePath, "metadata"), metadata); err != nil {
+		return err
+	}
+
+	if err := writeDataStore(filepath.Join(m.filePath, "snapshots"), snapshots); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *forkManagerImpl) initializeContractValidatorSet() (valset.ValidatorSet, error) {
+	return contract.NewContractValidatorSet(
+		m.logger,
+		m.blockchain,
+		m.executor,
+		m.GetSigner,
+		m.epochSize,
+	)
+}
+
+func (m *forkManagerImpl) registerPoAHooks(
+	hooks *hook.HookManager,
+	fork *IBFTFork,
+	height uint64,
+) error {
+	valSet, err := m.GetValidatorSet(height)
+	if err != nil {
+		return err
+	}
+
+	registerValidatorSetHook(hooks, valSet)
+
+	return nil
+}
+
+func (m *forkManagerImpl) registerPoAPrepareHooks(
+	hooks *hook.HookManager,
+	fork *IBFTFork,
```

Comment:
This param is not used

Reply:
Removed at 0ba08ab406b2499d88a22b9935d9555a012ff3e6